### ISSUE TITLE
Feature Request: PHPDoc generics typehints

### ISSUE THREAD

**Issue by tyteen4a03** [phpspec/prophecy#544]
It would be amazing if we could typehint our prophecies with Generics phpdoc:

```

/**
 * @var ObjectProphecy<ClassBeingMocked>
 */
private $prophecy;
```

**Comment by ciaranmcnulty** [comment]
This sounds like a psalm/phpstan/jetbrains feature request...

**Comment by tyteen4a03** [comment]
There needs to be `@template` annotations in place on Prophecy's side.

**Comment by stof** [comment]
For phpstan, you can use https://packagist.org/packages/jangregor/phpstan-prophecy which provides the stubs for the generics, and also extra extensions for better support of the magic calls.
Due to our signature based on magic calls (and the additional extra magic of `Argument::cetera()`), I don't think we can make our code statically analysable based only on phpdoc, without relying on a phpstan extension.

**Comment by gnutix** [comment]
But adding @template on ObjectProphecy would at the very least allow PHPStorm (and maybe other IDEs) to consider the types too, and not raise inspections everywhere, no ?

**Comment by stof** [comment]
With `@phpstan-this-out` being implemented in phpstan 1.9.0, it becomes possible to represent the behavior of `willImplement()` in phpdoc.
The phpstan extension will still be necessary to support our API based on magic calls.

**Comment by ciaranmcnulty** [comment]
I think we can/should import some subset of the types from those stubs, then users would get *some* benefit from them even without the extension (which probably would still be needed)
